test/store-user-intent.js

```diff
@@ -0,0 +1,52 @@
+'use strict';
+/* eslint max-len: 0 */
+
+const test = require('tape');
+const { parseArgs } = require('../index.js');
+
+
+// Rationale
+//
+// John Gee:
+// Looks like a boolean option, stored like a flag. Looks like a string option, stored like a string option.
+// No loss of information. No new pattern to learn in result.
+//
+// Jordan Harband: In other words, the way they're stored matches the intention of the user,
+// not the configurer, which will ensure the configurer can most accurately respond to the
+// user's intentions.
+
+test('when use string short option used as boolean then result as if boolean', (t) => {
+  const passedArgs = ['-o'];
+  const stringOptions = { opt: { short: 'o', type: 'string' } };
+  const booleanOptions = { opt: { short: 'o', type: 'boolean' } };
+
+  const stringConfigResult = parseArgs({ args: passedArgs, options: stringOptions, strict: false });
+  const booleanConfigResult = parseArgs({ args: passedArgs, options: booleanOptions, strict: false });
+
+  t.deepEqual(stringConfigResult, booleanConfigResult);
+  t.end();
+});
+
+test('when use string long option used as boolean then result as if boolean', (t) => {
+  const passedArgs = ['--opt'];
+  const stringOptions = { opt: { short: 'o', type: 'string' } };
+  const booleanOptions = { opt: { short: 'o', type: 'boolean' } };
+
+  const stringConfigResult = parseArgs({ args: passedArgs, options: stringOptions, strict: false });
+  const booleanConfigResult = parseArgs({ args: passedArgs, options: booleanOptions, strict: false });
+
+  t.deepEqual(stringConfigResult, booleanConfigResult);
+  t.end();
+});
+
+test('when use boolean long option used as string then result as if string', (t) => {
+  const passedArgs = ['--bool=OOPS'];
+  const stringOptions = { bool: { short: 'b', type: 'string' } };
+  const booleanOptions = { bool: { short: 'b', type: 'boolean' } };
+
+  const stringConfigResult = parseArgs({ args: passedArgs, options: stringOptions, strict: false });
+  const booleanConfigResult = parseArgs({ args: passedArgs, options: booleanOptions, strict: false });
```

Comment:
**thought (non-blocking):** I'm surprised by the results of the `boolean` assertion. If I the author explicitly say `--bool` is `type: 'boolean'`, I don't understand why we would capture values from incorrect usage by users.

```js
// node user-intent.js --bool=OOPS --more-flags...

const result = parseArgs({ options: { bool: { type: 'boolean' } } })

for (const [longOption, optionValue] of Object.entries(result.values)) {
  if (typeof optionValue === 'string') // process string options
  if (typeof optionValue === 'boolean') // process boolean options
}
```
☝️ This is an incredibly contrived example, but if my intent as the author was to do one thing with `string`s and another with `boolean`s it would break due to incorrect argument usage.

Note: This makes total sense to me if there was no configuration options, but if provided, I would think you're saying this is the argument contract my program accepts and relies on internally. If your program allows `boolean-like` options to accept values, you would simply set the `type` to `string` and account for that in your implementation. Thoughts?

Reply:
100% if i configure an arg to be a boolean, it should throw if it's not one.

Reply:
We have `strict:false` and `strict:true`.

In `strict:true` mode, the "wrong" usage will throw (well, hopefully, soon!). Note, these tests use `strict:false` in anticipation of that arriving.

In `strict:false` mode we are doing best effort parsing no matter the input. We don't want to throw away information which might be needed for the author to make use of. We hope they pay some attention...

Reply:
I understood the test was for `strict: false` and that we would like to capture user intent without throwing away information. However, that seems more applicable to the zero-config behavior to me. If I configure a CLI to accept specific input, I should be confident the args parser will return the type I defined.

[TS Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBA8mwEsD2A7AwqgZgg5lAvFAN4BQUUokA-AFxQDkARkkgDYQCGK9UAPgwGdgAJwQoc9MlAEALJMOC1pIsTikBbAK6tEYdkuZtOKEgF8SJStAwpsOOIlQCAPABlU9+MhRQIAD2AIFAATASgABWEkSAUQAGkIEAIIqJjQBJAAPmTScnIAbVcoMSh3cQdvAF06CtQbOzMLMUDhTA4AY2ssXGdalDD-QJCw+tw+gWzcvKFRdsU6EU0IKXIOYRwBJRnVfMqVqGjHfqVxxpJmiFaO6HC1gQhggEF1l3HfAKDQqFHPI4nifaYVgcDZ0KZ5PL5GDFHwAa0SSEwsC8TkqJxR-ShlXy9Cs9Eq7yGX3o23EPCoylE4igdEM7C4+3MEIAbhxWEsBGD9hCoTCoPCQIjkX80TUMQIsTi8QTBp8wiSVGSoBTSXhaSx6T5+KrGfswEgBAgjmzOZSdntzCRghB2sDhNBMJoUHNvFAwHcIM8Nr1xYS5d9ur9vBMABTtQNKH4+v6ZACUdFuwnuTxe0eDmQs4f6wDdHuCyXdSc9LxDU0Owa5EPImBYYIo4AgdAVVIkUFMABpuVBGGs61Ym3TjDwO12ewAvMFMttmWOZpw5mtIAt5gB0rPZEAEK8XUAA9LuoABabKqvhQJ3W7AoB4kLNCbtrZdF4Jrtkclc94R7g-Hs3U-iDlwZ4XhAV43neObjk+yavhuW5QfuR4noqeAARqxjASEoFiDeJCIQAckgiCdFAAAGn6kVAXD5uRHBjpRazQFmdiaPa+YcGEpH9t26FcKR7ZQCAEA5sAMjQPaAjaMAYTtEBjDQBwZGqqRK4kEAA)
I took a shot at implementing some TypeScript definitions for this and it's not feeling very intuitive imo.
<img width="760" alt="image" src="https://user-images.githubusercontent.com/32409546/157163837-e983b057-eb54-4eb8-9ee5-d45a13ba4f1b.png">

Notice `bar` and `baz` are configured as `type: boolean`, yet the results can be a `string`.

Reply:
Again this is non-blocking, but think this could lead to less than ideal implementations that require checking if `type: 'boolean'` options are actually `boolean`s. As well as convoluted TypeScript definitions that have different parse args results based on the `strict` mode config. Honestly, this isn't a major issue and really just poking at possible holes in the implementation as we get closer to the Node release.

Reply:
1) The implicit boolean and explicit boolean are not currently different. To make them behave differently adds another layer of complexity to the story (and implementation). Doable, but more complex.

2) What behaviour do you have in mind for mis-using configured types? (Throwing, despite being `strict:false`?)

I have been looking into some of the detail of what Minimist and Yargs actually do, and for interest Minimist returns empty string for a configured "string" option which is misused without a value. (This gives the type safety at the expense of obscuring the potential misuse!)

3) A high level question is why do you want to use `strict:false` mode at all? (The two use cases I have suggested are take-your-chances zero-config quick programs, and careful bolt-on behaviour.)

Reply:
Regarding TypeScript, you said:

> // Notice `bar` and `baz` are configured as `type: boolean`, yet the results can be a `string`.

The converse should apply to `foo` too, it is configured as `type: string`, yet the result can be `true`.

Reply:
> 2. What behavior do you have in mind for mis-using configured types? (Throwing, despite being strict:false?)

Not throw, I was thinking we just return a `boolean` regardless if it's misconfigured.
```js
// node misconfigured.js --foo=bar
const parsed = parseArgs({ strict: false, options: { foo: { type: 'boolean' } } })
parsed.values.foo // true
```

The input is already flawed, so might as well return the type I expect (as opposed to a completely different type `string`). Missing `--foo` would still be `undefined`.  Thoughts?

> and for interest Minimist returns empty string for a configured "string" option which is misused without a value

Not against this either tbh, but don't have strong opinions. I'm more "concerned" about `parseArgs` returning different types than the author defined. 

> 3. A high level question is why do you want to use `strict:false` mode at all?

I will likely always use `strict:true` and haven't put much thought into what use cases folks have for `strict:false`. Would be curious to hear your thoughts here!

Reply:
> The converse should apply to `foo` too, it is configured as `type: string`, yet the result can be `true`.

Oh really!? Maybe I do prefer the Minimist approach. i.e. setting an empty `string` for a misconfigured `type: string`

It would be great if authors didn't have to check the `typeof` before using prototype methods for a defined option type: e.g.
```js
// node misconfigured.js --foo
const parsed = parseArgs({ strict: false, options: { foo: { type: 'string' } } })

if (parsed.values.foo?.includes('substring')) // Crashes the program
```

Reply:
> It would be great if authors didn't have to check the typeof before using prototype methods for a defined option type

Without a throw, coercing the type means the program does not crash, but does not ensure safe and intended behaviour. In that sense crashing is arguably better behaviour than silently doing the wrong thing. Two examples:

```
// explode is configured as a boolean (but user mis-used)
$ node detonate.js --explode=false
```

```
// no-explode is configured as a boolean (but user typo)
$ node detonate.js --no-expode foo bar
```

Reply:
Great examples! Definitely demonstrates the severity and potential impact of misconfigured options. So I'm realizing `strict:false` is probably more suited for bring your own validation and making authors responsible for robustness. Whereas, `strict:true` allows authors to rely on `parseArgs` for option validation and throwing errors for misuse.